// @colyseus/schema/lib/encoding/spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPERATION = exports.TYPE_ID = exports.SWITCH_TO_STRUCTURE = void 0;
exports.SWITCH_TO_STRUCTURE = 255; // (decoding collides with DELETE_AND_ADD + fieldIndex = 63)
exports.TYPE_ID = 213;
/**
 * Encoding Schema field operations.
 */
var OPERATION;
(function (OPERATION) {
    OPERATION[OPERATION["ADD"] = 128] = "ADD";
    OPERATION[OPERATION["REPLACE"] = 0] = "REPLACE";
    OPERATION[OPERATION["DELETE"] = 64] = "DELETE";
    OPERATION[OPERATION["DELETE_AND_MOVE"] = 96] = "DELETE_AND_MOVE";
    OPERATION[OPERATION["MOVE_AND_ADD"] = 160] = "MOVE_AND_ADD";
    OPERATION[OPERATION["DELETE_AND_ADD"] = 192] = "DELETE_AND_ADD";
    /**
     * Collection operations
     */
    OPERATION[OPERATION["CLEAR"] = 10] = "CLEAR";
    /**
     * ArraySchema operations
     */
    OPERATION[OPERATION["REVERSE"] = 15] = "REVERSE";
    OPERATION[OPERATION["MOVE"] = 32] = "MOVE";
    OPERATION[OPERATION["DELETE_BY_REFID"] = 33] = "DELETE_BY_REFID";
    OPERATION[OPERATION["ADD_BY_REFID"] = 129] = "ADD_BY_REFID";
})(OPERATION || (exports.OPERATION = OPERATION = {}));
