// @colyseus/schema/lib/types/symbols.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.$fieldIndexesByViewTag = exports.$viewFieldIndexes = exports.$refTypeFieldIndexes = exports.$numFields = exports.$descriptors = exports.$onDecodeEnd = exports.$onEncodeEnd = exports.$childType = exports.$changes = exports.$deleteByIndex = exports.$getByIndex = exports.$filter = exports.$decoder = exports.$encoder = exports.$track = void 0;
exports.$track = Symbol("$track");
exports.$encoder = Symbol("$encoder");
exports.$decoder = Symbol("$decoder");
exports.$filter = Symbol("$filter");
exports.$getByIndex = Symbol("$getByIndex");
exports.$deleteByIndex = Symbol("$deleteByIndex");
/**
 * Used to hold ChangeTree instances whitin the structures
 */
exports.$changes = Symbol('$changes');
/**
 * Used to keep track of the type of the child elements of a collection
 * (MapSchema, ArraySchema, etc.)
 */
exports.$childType = Symbol('$childType');
/**
 * Optional "discard" method for custom types (ArraySchema)
 * (Discards changes for next serialization)
 */
exports.$onEncodeEnd = Symbol('$onEncodeEnd');
/**
 * When decoding, this method is called after the instance is fully decoded
 */
exports.$onDecodeEnd = Symbol("$onDecodeEnd");
/**
 * Metadata
 */
exports.$descriptors = Symbol("$descriptors");
exports.$numFields = "$__numFields";
exports.$refTypeFieldIndexes = "$__refTypeFieldIndexes";
exports.$viewFieldIndexes = "$__viewFieldIndexes";
exports.$fieldIndexesByViewTag = "$__fieldIndexesByViewTag";
